src/util/any_ref.h

```diff
@@ -0,0 +1,82 @@
+/*******************************************************************\
+
+Module: Any reference - type-safe wrapper around void*
+
+Author: Diffblue ltd.
+
+\*******************************************************************/
+
+#ifndef CPROVER_UTIL_ANY_REF_H
+#define CPROVER_UTIL_ANY_REF_H
+
+#include <type_traits>
+#include <typeinfo>
+
+/// Any reference
+/// Stores a reference to any type
+/// Call .at<T>() to get the reference back
+class any_reft final
+{
+public:
+  template<typename T>
+  explicit any_reft(T &val):
+    m_type(&type_of<T>),
+    m_ptr(const_cast<typename std::remove_const<T>::type *>(&val)) { }
+
+  template<typename T>
+  any_reft &operator=(T &val)
+  {
+    m_type=&type_of<T>;
+    m_ptr=const_cast<typename std::remove_const<T>::type *>(&val);
+    return *this;
+  }
+
+  template<typename T>
+  bool operator==(const T &val) const
+  {
+    return (
+      m_type==&type_of<T>
+      || m_type==&type_of<typename std::remove_const<T>::type>)
+        && *static_cast<T *>(m_ptr)==val;
+  }
+
+  /// Convert to a reference of a specific type
+  /// Throw std::bad_cast if type doesn't match the type
+  template<typename T>
+  T &as()
+  {
+    if(m_type==&type_of<T>
+      || m_type==&type_of<typename std::remove_const<T>::type>)
```

Comment:
|| at end of previous line (several occurrences)

Reply:
I'd argue that having a line starting with the operator is clearer.

New line starting with a letter suggests function call argument split.
```C++
if (func(argument1,
  argument2 ||
  argument3))
```
This one's more clear
```C++
if (func(argument1,
  argument2
  || argument3))
```

Becomes even more apparent with longer variable names:
```C++
if (func(anUnnecessarilyLongVariableName,
  anotherUnnecessarilyLongVariableName
  || aVariableNameThatStartsAndEndsInTwoDifferentTimezones))
```

Reply:
See point 2 in https://github.com/diffblue/cbmc/blob/develop/CODING_STANDARD.md